drv/lpc55-sha256/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    ///    SHA256.
+    ///
+    /// In both of these cases, wrapping is fine -- SHA256 actually specifies
+    /// the count as wrapping at 64 bits. So using a `Wrapping<u64>` saves some
+    /// overflow checks.
+    ///
+    /// This is `u64` instead of `usize` because you can call `update`
+    /// repeatedly with new slices, meaning the number stored here can easily
+    /// exceed `usize::MAX` on a 32-bit platform.
+    word_count: Wrapping<u64>,
+    /// Notification mask corresponding to the HASHCRYPT interrupt in whatever
+    /// code is hosting this driver. Used to sleep when things aren't ready.
+    notification_mask: u32,
+}
+
+impl<'a> Hasher<'a> {
+    /// Starts a new SHA256 hash operation, initializing the `HASHCRYPT` unit.
+    ///
+    /// You are _strongly advised_ to have reset the HASHCRYPT unit just before
+    /// this, lest it contain nonsense left over from a previous hash. (In
+    /// particular, the boot ROM leaves nonsense in it, so it needs to be reset
+    /// at least once after startup.)
+    ///
+    /// This driver doesn't reset the peripheral itself so that we don't make
+    /// assumptions about how reset is reached (e.g. by an IPC).
+    ///
+    /// `notification_mask` is the bitmask corresponding to the HASHCRYPT IRQ
+    /// routing to your task. If you'd rather busywait, pass 0.
+    #[inline(never)]
+    pub fn begin(
+        engine: &'a lpc55_pac::hashcrypt::RegisterBlock,
+        notification_mask: u32,
+    ) -> Self {
+        // Put the multi-function thingy into SHA2 256 mode and start a new
+        // hash. (The UM is not entirely clear whether setting the mode and
+        // starting the hash in a single write is legal, but it works and NXP's
+        // code appears to do the same thing.)
+        engine
+            .ctrl
+            .write(|w| w.mode().sha2_256().new_hash().start());
+
+        Self {
+            engine,
+            word_count: Wrapping(0),
+            notification_mask,
+        }
+    }
+
+    /// Extends the current hash-in-progress with the given `data`,
+    /// exclusive-ORing each word with `mask.
+    ///
+    /// `data` may cross block boundaries, be a partial block, etc. It will be
+    /// concatenated with the `data` passed to any other `update` call.
+    ///
+    /// In most cases you want a `mask` of 0, the parameter is provided because
+    /// it's useful in certain HMAC operations and using the same routine for
+    /// both cases saves some space.
+    #[inline(never)]
+    pub fn update(&mut self, data: &[u32], mask: u32) {
+        for &word in data {
+            self.load_word(word, mask);
+        }
+    }
+
+    /// Completes the SHA256 hash.
+    ///
+    /// You should seriously consider flipping the HASHCRYPT unit back into
+    /// reset after calling this, to avoid leaking data about
+    /// whatever-it-was-you-were-just-doing. Or don't; we're not the cops.
+    #[inline(never)]
+    pub fn finish(mut self) -> [u32; 8] {
```

Comment:
Would it be correct to use `WORDS_PER_HASH` here?  It has the right value, but not sure if it's got the correct _meaning_.
```suggestion
    pub fn finish(mut self) -> [u32; WORDS_PER_HASH] {
```